### PR TITLE
Initialise covid-19 data extract-transform

### PR DESCRIPTION
**Pull request by juliaogris**
Initialise covid19 data extract-transform in backend/main.go. Parse an
input CSV file from stdin and write human readable JSON of compacted data,
without Latitude Longitude and date repetitions to stdout in covid19
binary.

Create sample input backend/testdata/data.csv and expected output in
backend/testdata/want.json.

Add documentation in README.md and the usual go project scaffold: go.mod,
go.sum, Makefile, .gitignore, .golangci.yaml.

Add LICENSE and CI setup.

### CODE REVIEW

File: backend/main.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	return dates, nil
+}
+
+func parseCountries(rows [][]string, dates []time.Time) (map[string]*Country, error) {
+	countries := map[string]*Country{}
+	dateIdx := 0
+	dateCnt := len(dates)
+	for _, row := range rows {
+		country, province := row[1], row[2]
+		if province != "" {
```
Comment (camh-) [inline backend/main.go:98]: This looks like it skips countries with provinces. At least for Australia it does - there are no Australian data with an empty province in the CSV. Same with Canada too.

Comment (juliaogris) [reply backend/main.go]: yes, shame on me for being so sloppy.
i have addressed this.

File: backend/main.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+// specific country some meta-data such as population.
+type Country struct {
+	Name       string  `json:"name"`
+	Lat        float64 `json:"lat"`
+	Lng        float64 `json:"lng"`
+	Population int64   `json:"population"`
+	// per date new, not cumulative, confirmed cases/recovered/dead:
+	Confirmed []int `json:"confirmed"`
+	Recovered []int `json:"recovered"`
+	Dead      []int `json:"dead"`
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply backend/main.go:38]: don't like my Austrian directness anymore?

[the remark this answers is not in the document]
Comment (camh-) [reply backend/main.go:38]: Maybe, sort of, i mean i'm not saying i like it and i'm not saying i dont like it, but everything has its place.

File: backend/main.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	}
+	country.Confirmed[dateIdx] += confirmed
+	country.Recovered[dateIdx] += recovered
+	country.Dead[dateIdx] += dead
+	date, err := time.Parse("2006-01-02", row[0])
+	if err != nil {
+		return errs.Errorf("%v: %s, %s: %v", ErrInvalidInput, country.Name, row[0], err)
+	}
+	if wantDate != date {
+		return errs.Errorf("%v: %s: want date %s got %s", ErrInvalidInput, country.Name, row[0], wantDate, date)
```
[the remark this answers is not in the document]
Comment (camh-) [reply backend/main.go]: row row row your boat

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+cover: test  ## Show test coverage in your browser
+	go tool cover -html=$(COVERFILE)
+
+CHECK_COVERAGE = awk -F '[ \t%]+' '/^total:/ {print; if ($$3 < $(COVERAGE)) exit 1}'
+FAIL_COVERAGE = { echo '$(COLOUR_RED)FAIL - Coverage below $(COVERAGE)%$(COLOUR_NORMAL)'; exit 1; }
+
+.PHONY: test check-coverage cover
+
+# --- Lint ---------------------------------------------------------------------
+GOLINT_VERSION = 1.28.3
```
Comment (camh-) [inline Makefile:44]: there is a 1.30.0 version now. dont mind if you want to stay at 1.28.3 for reasons, just pointing it out.

File: backend/main_test.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func TestParse(t *testing.T) {
+	s := `
+Date, 	    Coutry, Province, Lat, Long, Confirmed, Recovered, Deaths
+2020-01-01,    Abc,         ,   2,    1,         3,         0,      1
+2020-01-02,    Abc,         ,   2,    1,        10,         2,      6
+2020-01-03,    Abc,         ,   2,    1,        12,         4,      5
+`
+	b := bytes.NewBufferString(s[1:])
```
Comment (camh-) [inline backend/main_test.go:29]: I wonder if `s[1:]` works on windows which uses CR,LF as line terminators? Perhaps trim it or find the first non-whitespace? I wonder if you even care about windows?

File: backend/main_test.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	want := &Data{
+		Dates: []time.Time{mustTime("2020-01-01"), mustTime("2020-01-02"), mustTime("2020-01-03")},
+		Countries: map[string]*Country{
+			"Abc": {
+				Name:      "Abc",
+				Lat:       2,
+				Lng:       1,
+				Confirmed: []int{3, 7, 2},
+				Recovered: []int{0, 2, 2},
+				Dead:      []int{1, 5, -1},
```
Comment (camh-) [inline backend/main_test.go:42]: If this is possible, I think the "country" needs to be Golgotha.

File: backend/main_test.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+	t, err := time.Parse("2006-01-02", s)
+	if err != nil {
+		panic(err)
+	}
+	return t
+}
+
+func TestParse(t *testing.T) {
+	s := `
+Date, 	    Coutry, Province, Lat, Long, Confirmed, Recovered, Deaths
```
Comment (camh-) [inline backend/main_test.go:24]: this has a tab in it, which makes it not line up on my screen, which uses 8 spaces for tabs as everything should.

Also, Coutry is a mispelling of Country

File: backend/main_test.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+				Dead:      []int{1, 5, -1},
+			},
+		},
+	}
+	require.Equal(t, want, got)
+}
+
+func TestParseProvince(t *testing.T) {
+	s := `
+Date, 	    Coutry, Province, Lat, Long, Confirmed, Recovered, Deaths
```
Comment (camh-) [inline backend/main_test.go:51]: same problem with this heading as the last

File: backend/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (juliaogris) [reply backend/main.go]: yeah ok, sorry. updated for all with `row` - more useful for debugging.

File: backend/main_test.go (hunk 6 above)
Comment (juliaogris) [reply backend/main_test.go:42]: or just a country where people do shitty jobs with data entry.

File: backend/main_test.go (hunk 5 above)
Comment (juliaogris) [reply backend/main_test.go:29]: you are right-  i don't care about windof (as we lovingly call it in german -  doof means dumb).
changed to TrimSpace

File: backend/main_test.go (hunk 6 above)
Comment (juliaogris) [reply backend/main_test.go:42]: so pretty much could be any.

File: backend/main_test.go (hunk 7 above)
Comment (juliaogris) [reply backend/main_test.go:24]: sorry, fixed.

File: backend/main_test.go (hunk 8 above)
Comment (juliaogris) [reply backend/main_test.go:51]: done. thanks.

File: Makefile (hunk 4 above)
Comment (juliaogris) [reply Makefile]: done.

**Pull request merged**